### ISSUE TITLE
e-learning

### ISSUE THREAD

**Issue by spametki** [rochacbruno-archive/Movuca#10]
<h5>As suggested by Bruno, here is a new discussion on the plans for adding e-learning support for Movuca project:</h5>

This discussion was started and even might continue in the web2py users list:
http://groups.google.com/group/web2py/browse_thread/thread/428225259aafd360/34280fa8641cce66

Some extracts of the web2py list discussion:
- "...I need to improve the content-types system to be able to
  include content-bundles, a content bundle is like a folder, a group, a
  course, a class, a complete tutorial or whatever you need to group
  together. With a generic functionality like that I think it will be easy to
  implement it as a courseware, also implementing better privacy options
  based on groups, permissions and users who have signed to some course or
  "bundle".  ..."
- "...For the videos I am thinking in integration with Big Blue Button API, VIMEO
  Plus API, Amazon S3 for streaming, and also Google Docs. 
  ..."

<h5>The proposed tasks (Bruno)</h5>

<em>Basically I think this are the first steps:
- \- Create the desired content types, datamodel, code, views and index
  component.*
  Video
  Tutorial
  LEcture
  HowTo
  Discussion
  etc..
- \- Create a "Content-Bundle" feature.*
  So you can click at "create a new bundle" choose a name, privacy settings,
  studend-group, price and other configs
  After that, you can insert new contents in to a "bundle", it can be a
  "programming course level 1" and inside that bundle you can insert videos,
  tutorials, lectures, discussions.... etc..
- \- Create a "group/community" functionality*
- \- Create a general forum*

Second steps maybe:
- \- Create an admin dashboard*
  this dashboard should expose different workspaces for users, publishers,
  editors, moderators and administrators.
  it has to be based on web2py Auth (group, membership etc)</em>

<h5>Note</h5>

There's one more thing: I belive that the Movuca app needs some sort of introductory document to understanding the application design for developers. Movuca doesn't follow the standard structure of a simple MVC web2py application and could be difficult for colaborators (i.e. me) to submit modifications coherent with the project's architecture. Perhaps there is such documentation and I didn't find it?.

Alan Etkin

**Comment by spametki** [comment]
Some classes not listed in Bruno's roadmap:
(it's just a draft of course)
- Test:
  Object for storing different kinds of  exam tasks
  - quizes (sequence of Quiz objects)
  - tasks (sequence of assigned tasks)
  - starts (time)
  - due (time)
  - result (calculated or manual)
  - students (a sequence of one or more students)
  - faculty (the teachers assigned for helping)
  - evaluators (the teachers assigned for evaluation)
- Task:
  A work assigned to the student
  - description (text or article)
  - work (a field for uploading or link to the finished task)
  - points (for accomplishment)
- Quiz:
  A quiz can be a classic
  quiz or any evaluation feedback
  (including traditional no multiple choice
  questions that need redaction)
  unique quiz instance
  there's only one per user/team
  it's supposed to be created
  randomly
  (for evaluation and record purposes)
  - multi fk Stages (QuizStage object)
  - presentation (introductory article)
- QuizStage:
  quiz instance
  tentative properties
  - fk Question`*`
  - multi fk Options (Answer object)`*`
  - fk Correct (Answer object)`*`
  - tags (will be used for auto quiz shuffle)
  - points (right answer value)`***`
  - text answer (to be read before evaluation)`**`

`*` There are Answer objects in the app already
`**` If no options it is assumed that a text answer is required
`***` The value to compute in case of correct. It should be assigned to the external pre-existing model instance in a per-user basis (perhaps with an intermediate table)

**Comment by ndegroot** [comment]
Maybe you have found it already, there is an article on the demo site about creating a new content type see  http://movu.ca/demo/article/show/1/how-to-create-content-types-in-movuca-cms

I now starting to explore Movuca and will try to port some earlier work  (tests, mediasite player) into it. What do you think is a good way of working:
o create forks to test, develop enhancements at github
o create a 'central' elo-version of mocuva were we can add elo-modules and plugins, and test them and contribute fixes, and propose enhancements and plugins  back to movuca to be included?

I can host this version if you want and keep it in sync with the base version.

**Comment by spametki** [comment]
Thanks for the link. I wasn't aware of it. II think that the way of adding features should be decided by Bruno, which knows the app better than us. For me, I prefer a central project with patch requests, so there is a common filter for enhancements and the code is somewhat minimized, I get confused with paralell projects. And once the new features design is to be defined we can make some sort of distribution of the tasks.

**Comment by spametki** [comment]
There's something else. I belive that Bruno's goal is to add elearning as a built-in feature of Movuca, instead of a specific elearning Movuca version. If this is so, I agree since, there could be a common starting resource for further specialized developments (as I've seen is the case with Moodle)

**Comment by ndegroot** [comment]
Agreed that having a central codebase is the best solution. To make myself clear, I don't propose a diverging separate version of Movuca, just a  'playground' and maybe a location for optional extra's.  Say I develop a content-type or a mocuva-dependant plug-in, but it's not convenient to add it to Mocuva for valid reasons, but still useful for elo-use.

**Comment by spametki** [comment]
Alright, I misunderstood your point of view. Anyway, I am prototyping a mini web2py elearning app as a way of researching for simple features. When I have something relevant, I'll post the source in case there is something useful.

**Comment by rochacbruno** [comment]
Yes you are right, I want e-learning as a built-in, but I think the Movuca's goal is to be generic, to be used with any kind of social websites.

So you can have e-learning features as generic content types, it is free to change anything in the core if needed.

**Comment by spametki** [comment]
My e-learning development work was suspended due to my customer choice for Moodle. Perhaps my Python and web2py advocacy was not good enough. In Argentina, PHP seems to be the ruler for this kind of cases.
However, I still am interested in devolping the feature, though with a much more paced attention.

I was working in an e-learning tool for web2py that is now in an initial development phase. The project is open source and is hosted here:

http://code.google.com/p/pyodel/

It aims to provide a common set of e-learning tools pluggable with any web2py CMS (wether it is Instant Press, Movuca, or other).
